+Base/QRconv.m

```diff
@@ -0,0 +1,384 @@
+    
+    for ii = 1:size(V,2)
+        dduplicates = V(1,:) == V(1, ii) & V(2,:) == V(2, ii);
+        
+        if sum(dduplicates) > 1
+            duplicates = duplicates | dduplicates;
+        end
+    end
+    
+    if isfield(options_guess, 'X_expected') && isfield(options_guess, 'Y_expected')
+        V_expected = [options_guess.X_expected; options_guess.Y_expected];
+    else
+        V_expected = [NaN; NaN];
+    end
+    b_expected = [options_guess.img_W; options_guess.img_H]/2 - M_guess * V_expected;
+    
+    % Setup variables that we will change as we loop.
+    mostvotes = 1;
+    b_guess = [NaN; NaN];
+    outliers = true(1, size(b_guesses,2));
+    dist = 3*options_guess.d;
+    
+    % Radius within which b guesses are considered the same guess.
+    R = options_guess.d;
+    
+    for ii = 1:size(b_guesses,2)                                    % For every candidate...
+        if ~any(isnan(b_guesses(:, ii))) && ~duplicates(ii)
+            votes = sum((b_guesses - b_guesses(:, ii)).^2) < R*R & ~duplicates;   % How many other candidates agree?
+            
+            if mostvotes <= sum(votes)                              % If this is a new record...
+                b_guess = mean(b_guesses(:, votes), 2);             % Estimate b as the average.
+                dist_ = norm(b_guess - b_expected);
+                
+                if (sum(votes) > 0 && dist_ < dist) || (mostvotes < sum(votes) || dist_ < dist)
+                    dist = dist_;
+                    mostvotes = sum(votes);                         % Record the record.
+                    outliers = ~votes;                              % Record the candidates that were outside.
+                end
+            end
+        end
+    end
+    
+%     if sum(~outliers) < 1 && ~any(isnan(b_expected)) && norm(b_guess - b_expected) > 3*options_guess.d
+%         outliers(:) = true;
+%     end
+    
+    % Trim the outliers.
+    v_trim = v(:, ~outliers);
+    V_trim = V(:, ~outliers);
+
+    if isempty(v_trim)
+        M = [[NaN NaN]; [NaN NaN]];
+        b = [NaN; NaN];
+        
+        M2 = M;
+        b2 = b;
+        
+        return;
+    end
+    
+    M2 = M_guess;
+    b2 = b_guess;
+
+    if sum(~outliers) >= 2
+        % Fit the candidates fully to an affine transformation.
+        fun = @(p)( leastsquares(v_trim, V_trim, [p(1:2)', p(3:4)'], p(5:6)') );
+        p_guess = [M_guess(:); b_guess]';
+        p_full = fminsearch(fun, p_guess, struct('TolFun', 1, 'TolX', 1e-1));
+
+        M = [p_full(1:2)', p_full(3:4)'];
+        b = p_full(5:6)';
+    else
+        M = M2;
+        b = b2;
+    end
+end
+
+function img = threshold(img)
+%     img = imbinarize(img);
+    img = imbinarize(imgaussfilt(img,1));
+end
+function img = flatten(img)
+%     img = imgaussfilt(img,10) - imgaussfilt(img,1);
+% class(imgaussfilt(img,10))
+%     img = imgaussfilt(img,10) - img;
+    img = imgaussfilt(img,10) - imgaussfilt(img,3);
```

Comment:
Reason these values are hardcoded? Can we tie it to something about the image, pixels or something?

Reply:
They shouldn't be hardcoded. A future enhancement should open them up to the user. It might be a good task for @cassie-song in a future pull request.